core/tee/tee_cryp_utl.c

```diff
@@ -329,34 +314,64 @@ TEE_Result tee_aes_cbc_cts_update(void *cbc_ctx, void *ecb_ctx,
 	return TEE_SUCCESS;
 }
 
-TEE_Result tee_prng_add_entropy(const uint8_t *in, size_t len)
-{
-	return crypto_rng_add_entropy(in, len);
-}
-
 /*
  * Override this in your platform code to feed the PRNG platform-specific
  * jitter entropy. This implementation does not efficiently deliver entropy
  * and is here for backwards-compatibility.
  */
-__weak void plat_prng_add_jitter_entropy(void)
+__weak void plat_prng_add_jitter_entropy(enum crypto_rng_src sid,
+					 unsigned int *pnum)
 {
 	TEE_Time current;
 
+#ifdef CFG_SECURE_TIME_SOURCE_REE
+	if (CRYPTO_RNG_SRC_IS_QUICK(sid))
+		return; /* Can't read REE time here */
+#endif
+
 	if (tee_time_get_sys_time(&current) == TEE_SUCCESS)
-		tee_prng_add_entropy((uint8_t *)&current, sizeof(current));
+		crypto_rng_add_event(sid, pnum, &current, sizeof(current));
 }
 
-__weak void plat_prng_add_jitter_entropy_norpc(void)
+__weak void plat_rng_init(void)
 {
+	TEE_Result res = TEE_SUCCESS;
+	TEE_Time t;
+
 #ifndef CFG_SECURE_TIME_SOURCE_REE
-	plat_prng_add_jitter_entropy();
+	/*
+	 * This isn't much of a seed. Ideally we should either get a seed from
+	 * a hardware RNG or from a previously saved seed.
+	 *
+	 * Seeding with hardware RNG is currently up to the platform to
+	 * override this function.
+	 *
+	 * Seeding with a saved seed will require cooperation from normal
+	 * world, this is still TODO.
+	 */
+	res = tee_time_get_sys_time(&t);
+#else
+	memset(&t, 0, sizeof(t));
```

Comment:
This case is obviously even worse than above. Maybe `fortuna_read()` should print a big fat warning if it's called before any more entropy is added?

Reply:
The Fortuna implementation doesn't keep track of quality of the entropy so there's no easy way of doing a print for that later. How about:
```
EMSG("Warning: seeding RNG with zeroes");
```
Right here?

Reply:
Yep sounds good!